[motion-path] Check if we are mid layout when setting containing block rect for ray

nmoucht opened the pull request:
#### acfa97d238a39af164eaf24a0abd4eec0cf096b4
<pre>
[motion-path] Check if we are mid layout when setting containing block rect for ray
<a href="https://bugs.webkit.org/show_bug.cgi?id=260110">https://bugs.webkit.org/show_bug.cgi?id=260110</a>
rdar://113780201

Reviewed by Simon Fraser.

On the first RenderLayer::updateTransform call the parent is mid layout, so it is
inorrect to query it about its rect size. If we have a motion path transform, set
that this RenderObject needs a transform update on the parent through the FrameView&apos;s
layout context, and have the parent call updateTransform after it has completed layout.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-019-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-019.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-020-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-020.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-021-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-021.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-022-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-022.html: Added.

Canonical link: <a href="https://commits.webkit.org/267479@main">https://commits.webkit.org/267479@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/653d25c8b2ded9740a9d3d839ec5b92e4ba1a73e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16704 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17028 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18486 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15653 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16894 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20260 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17167 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16903 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17279 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14453 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19271 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14523 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15137 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15512 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15303 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19601 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15906 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15090 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/14977 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19456 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2063 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15737 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderBlockFlow.h:
```diff
@@ -551,6 +551,8 @@ class RenderBlockFlow : public RenderBlock {
     // FIXME: This is temporary until after we remove the forced "line layout codepath" invalidation.
     std::optional<LayoutUnit> m_previousModernLineLayoutContentBoxLogicalHeight;
 
+    Vector<RenderLayer*> m_childLayersNeedingTransformUpdate;
```
This violates our "no new raw pointers" rule. This needs to use weak pointers.

nmoucht replied:
Right fixed.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1335,29 +1335,35 @@ void RenderLayer::updateTransformFromStyle(TransformationMatrix& transform, cons
+
+            // If the containing block has an incomplete rect this means it is mid layout. Have the parent re-run this
+            // after completing layout
+            if (!pathReferenceBoxRect.width() || !pathReferenceBoxRect.height()) {
```
Just checking for zero width or height seems wrong. What if it was laid out once before, but the size is changing?

nmoucht replied:
Right, we should check if the parent needs layout.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1335,29 +1335,35 @@ void RenderLayer::updateTransformFromStyle(TransformationMatrix& transform, cons
+            // after completing layout
+            if (!pathReferenceBoxRect.width() || !pathReferenceBoxRect.height()) {
+                if (auto* containingBlockFlow = dynamicDowncast<RenderBlockFlow>(containingBlock)) {
+                    containingBlockFlow->markNeedingTransformUpdate(this);
```
It's not ideal for render objects (containingBlockFlow) to have a cache of RenderLayers; RenderLayers are "downstream" from the render tree.

It's probably not ideal for this motion path code to live in RenderLayer; it's layout-related, so it should really live on some render object.

nmoucht replied:
Moved these bits to RenderBlockFlow

smfr commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -291,6 +291,35 @@ void RenderLayerModelObject::updateLayerTransform()
+    if (is<BoxPathOperation>(pathOperation)) {
+        auto& boxPathOperation = downcast<BoxPathOperation>(*pathOperation);
+        auto pathReferenceBoxRect = snapRectToDevicePixelsIfNeeded(referenceBoxRect(boxPathOperation.referenceBox()), *this);
+        boxPathOperation.setPathForReferenceRect(FloatRoundedRect { pathReferenceBoxRect });
```
You could `return` here and remove the `else`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -291,6 +291,35 @@ void RenderLayerModelObject::updateLayerTransform()
         layer()->updateTransform();
 }
 
+bool RenderLayerModelObject::setReferenceBoxForPathOperations()
```
It would be nice to make a new file with a helper class that knows about Motion Path stuff, rather than polluting RenderLayerModelObject with it.

nmoucht replied:
Added MotionPathHelper.

nmoucht replied:
We can factor out some of the other motion path code to this file in some follow up patches as well.

smfr commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -291,6 +291,35 @@ void RenderLayerModelObject::updateLayerTransform()
+            auto& rayPathOperation = downcast<RayPathOperation>(*pathOperation);
+            auto pathReferenceBoxRect = snapRectToDevicePixelsIfNeeded(parentBlock->transformReferenceBoxRect(parentBlock->style()), *this);
+            if (auto* containingBlockFlow = dynamicDowncast<RenderBlockFlow>(parentBlock)) {
+                if (parentBlock->needsLayout() && !containingBlockFlow->wasMarkedNeedingTransformUpdate(this)) {
```
It's a bit gross that this sets some state on the `containingBlockFlow`. Why not just change the code to always update motion path things after the containing block has finished layout?

nmoucht replied:
Changed to set this state in RenderLayer::updateTransform and moved the rest to MotionPathHelper.

nmoucht replied:
Not sure if `RenderLayer::updateTransform()` is the right place to set this state. We can either set it here or we have to set in every `RenderObject` subclass where we call `updateLayerTransform()` in the individual `layout()` functions.

nmoucht replied:
Moved the marking of needing a transform update to `RenderLayerModelObject::updateLayerTransform`.

smfr commented on Source/WebCore/rendering/RenderBlockFlow.h:
```diff
@@ -551,6 +551,8 @@ class RenderBlockFlow : public RenderBlock {
     // FIXME: This is temporary until after we remove the forced "line layout codepath" invalidation.
     std::optional<LayoutUnit> m_previousModernLineLayoutContentBoxLogicalHeight;
 
+    Vector<WeakPtr<RenderLayerModelObject>> m_descendantsNeedingTransformUpdate;
```
This is bloating RenderBlockFlow (of which there are potentially thousands on a web page) with a data member whose lifetime is transient; this member is only used during layout, so it should state on the stack or some per-layout state object.

nmoucht replied:
Moved to LocalFrameViewLayoutContext with a map of RenderBlock to the list of descendants needing updates. This isn't ideal but we can't use the RenderLayoutState stack since we would need the containing block's state not the current state (although maybe its ok to add a method that searches through the stack for the parent's state).

smfr commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -600,6 +600,10 @@ void RenderBlockFlow::layoutBlock(bool relayoutChildren, LayoutUnit pageLogicalH
         setPageLogicalOffset(state->pageLogicalOffset(this, logicalTop()));
 
+    for (auto& descendant : m_descendantsNeedingTransformUpdate)
+        descendant->updateLayerTransform();
```
This might have to happen before we compute layout overflow, so that the motion path-related transform on the child contributes to overflow on the container.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/MotionPathHelper.cpp:
```diff
@@ -0,0 +1,54 @@
+        auto& boxPathOperation = downcast<BoxPathOperation>(*pathOperation);
+        auto pathReferenceBoxRect = snapRectToDevicePixelsIfNeeded(renderer.referenceBoxRect(boxPathOperation.referenceBox()), renderer);
+        boxPathOperation.setPathForReferenceRect(FloatRoundedRect { pathReferenceBoxRect });
+    } else if (is<RayPathOperation>(pathOperation)) {
```
You can early return before this `if`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/MotionPathHelper.h:
```diff
@@ -0,0 +1,39 @@
+
+class RenderElement;
+
+class MotionPathHelper {
```
Not sure that "Helper" adds much. Maybe just `MotionPath` as the class and file name.

smfr replied:
Maybe the code in `RenderStyle::applyMotionPathTransform()` should move into this file as well.

nmoucht replied:
Fixed. Also moved applyMotionPathTransform to this file.

smfr commented on Source/WebCore/rendering/MotionPathHelper.cpp:
```diff
@@ -0,0 +1,54 @@
+            rayPathOperation.setContainingBlockReferenceRect(pathReferenceBoxRect);
+            auto left = renderer.style().left();
+            auto top = renderer.style().top();
+            rayPathOperation.setStartingPosition(FloatPoint(left.isPercent() ? left.value() / 100 * pathReferenceBoxRect.width() : left.value(), top.isPercent() ? top.value() / 100 * pathReferenceBoxRect.height() : top.value()));
```
Thinking about this a bit more, it's not good that we're pushing state backing into the PathOperation that's owned by `RenderStyle`. That is going to break style sharing, if elements with different sized containing blocks share style. Have you tested this?

I think we need to have the RenderStyle stuff (i.e. PathOperation) to be immutable, and store per-renderer state somewhere else. Is there a reason we can't just do this computation in `RenderStyle::applyMotionPathTransform()` ?

nmoucht replied:
I'm not convinced that this case would be broken since I think for each case we would set the appropriate state when doing updateTransform (will add a test for this). However, I agree its not right to keep this state on `PathOperation`, but since we cannot directly access this state in `applyMotionPathTransform` (which is why we had to use this solution in the first place), I refactored the code to pass the necessary information down into `applyMotionPathTransform`.

nmoucht replied:
Added a test case for this.

smfr commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -286,6 +287,11 @@ TransformationMatrix* RenderLayerModelObject::layerTransform() const
+    if (style().offsetPath()) {
+        MotionPathHelper::setReferenceBoxForPathOperations(*this);
+        if (auto* cb = containingBlock())
+            view().frameView().layoutContext().markNeedingTransformUpdate(cb, this);
```
You only need to call this for ray operations, right, so it's a waste to do this work for every renderer with offset path.

smfr replied:
Also you're going to go ahead and call updateTransform() here anyway, and then a second time later for things with ray operations?

nmoucht replied:
Right, added a function on `MotionPath` to specify if the `PathOperation` has a dependency on the containing block's size. We return early for this case so we don't calculate the transform twice.

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -660,6 +660,29 @@ void LocalFrameViewLayoutContext::popLayoutState()
+
+Vector<WeakPtr<RenderLayerModelObject>> LocalFrameViewLayoutContext::descendantsNeedingTransformUpdate(RenderLayerModelObject* parent)
+{
+    if (m_descendantsNeedingTransformUpdate.contains(parent))
```
This is two hash lookups. You could do it in one with `find()`, or maybe even `get()` does the right thing if there's no matching entry.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -660,6 +660,29 @@ void LocalFrameViewLayoutContext::popLayoutState()
+
+void LocalFrameViewLayoutContext::markFinishedLayout(RenderLayerModelObject* parent)
+{
+    if (m_descendantsNeedingTransformUpdate.contains(parent))
```
This is two hash lookups. OK to call `remove` without a check.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -660,6 +660,29 @@ void LocalFrameViewLayoutContext::popLayoutState()
 
+void LocalFrameViewLayoutContext::markNeedingTransformUpdate(RenderLayerModelObject* parent, RenderLayerModelObject* descendant)
+{
+    if (m_descendantsNeedingTransformUpdate.contains(parent)) {
```
`contains/set` can be done at the same time with `ensure()`.

nmoucht replied:
We can't do this since we need to check if the descendant is in the map (to prevent adding it again after this gets called when the containing blocks layout finishes).

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -660,6 +660,29 @@ void LocalFrameViewLayoutContext::popLayoutState()
+{
+    if (m_descendantsNeedingTransformUpdate.contains(parent)) {
+        auto descendants = m_descendantsNeedingTransformUpdate.get(parent);
+        if (!descendants.contains(descendant))
```
This is linear search.

nmoucht replied:
Needed due to above.

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.h:
```diff
@@ -192,6 +197,7 @@ class LocalFrameViewLayoutContext {
     std::unique_ptr<Layout::LayoutTree> m_layoutTree;
     std::unique_ptr<Layout::LayoutState> m_layoutState;
     std::unique_ptr<UpdateScrollInfoAfterLayoutTransaction> m_updateScrollInfoAfterLayoutTransaction;
+    HashMap<const RenderLayerModelObject*, Vector<WeakPtr<RenderLayerModelObject>>> m_descendantsNeedingTransformUpdate;
```
`WeakHashMap` ?

smfr replied:
The "descendants" in the name is confusing; here s a member variable. Would be better as `m_containersWith...`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -593,6 +593,9 @@ void RenderBlockFlow::layoutBlock(bool relayoutChildren, LayoutUnit pageLogicalH
         layoutPositionedObjects(relayoutChildren || isDocumentElementRenderer());
     }
+
+    updateDescendantTransformsAfterLayout();
```
Are there classes other than RenderBlockFlow that need this?

nmoucht replied:
Yeah I think the other subclasses of RenderBlock need this. Will add some tests on grid etc.

nmoucht replied:
Added to RenderBlock subclasses and added tests for grid and flex.

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.h:
```diff
@@ -128,6 +129,10 @@ class LocalFrameViewLayoutContext {
 
+    void markNeedingTransformUpdate(RenderLayerModelObject* parent, RenderLayerModelObject* descendant);
+    Vector<WeakPtr<RenderLayerModelObject>> descendantsNeedingTransformUpdate(RenderLayerModelObject* parent);
+    void markFinishedLayout(RenderLayerModelObject* parent);
```
Not a fan of `markFinishedLayout` (layout is not really finished for this block). It's more like `hasFinalSize()` or something.

nmoucht replied:
Changed to `finishedUpdatingDescendants`.

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.h:
```diff
@@ -128,6 +129,10 @@ class LocalFrameViewLayoutContext {
     UpdateScrollInfoAfterLayoutTransaction& updateScrollInfoAfterLayoutTransaction();
     UpdateScrollInfoAfterLayoutTransaction* updateScrollInfoAfterLayoutTransactionIfExists() { return m_updateScrollInfoAfterLayoutTransaction.get(); }
 
+    void markNeedingTransformUpdate(RenderLayerModelObject* parent, RenderLayerModelObject* descendant);
```
`parent` isn't really accurate; it's about containing block.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -3548,4 +3548,14 @@ LayoutUnit RenderBlock::layoutOverflowLogicalBottom(const RenderBlock& renderer)
 
+void RenderBlock::updateDescendantTransformsAfterLayout()
+{
+    auto descendants = view().frameView().layoutContext().descendantsNeedingTransformUpdate(this);
```
Rather than fetching `descendantsNeedingTransformUpdate` and then clearing later via `markFinishedLayout`, you could just have a single `take` function that uses `std::exchange()` to clear the hash map and return it.

nmoucht replied:
Needed due to above.

smfr replied:
It could still be `takeBoxesNeedingTransformUpdateAfterContainerLayout()` which would extract them from the hash map, and remove the need for `finishedUpdatingDescendants()`.

nmoucht replied:
This was originally to prevent us from erroneously adding the `RenderBox` back into the vector when we call `updateTransform` from in `updateDescendantTransformsAfterLayout`. However, we can get around this by directly calling `updateTransform` on the layer, bypassing the code in `RenderLayerModelObject::updateLayerTransform` that would incorrectly re-add it. This helps simplify the code by using take and removing `finishedUpdatingDescendants`.

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.h:
```diff
@@ -127,6 +128,9 @@ class LocalFrameViewLayoutContext {
 
     UpdateScrollInfoAfterLayoutTransaction& updateScrollInfoAfterLayoutTransaction();
     UpdateScrollInfoAfterLayoutTransaction* updateScrollInfoAfterLayoutTransactionIfExists() { return m_updateScrollInfoAfterLayoutTransaction.get(); }
+    bool markNeedingTransformUpdate(RenderLayerModelObject*, RenderLayerModelObject*);
```
Let's write this as
`bool setBoxNeedsTransformUpdateAfterContainerLayout(RenderBox&, RenderBlock& container);`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.h:
```diff
@@ -127,6 +128,9 @@ class LocalFrameViewLayoutContext {
     UpdateScrollInfoAfterLayoutTransaction& updateScrollInfoAfterLayoutTransaction();
     UpdateScrollInfoAfterLayoutTransaction* updateScrollInfoAfterLayoutTransactionIfExists() { return m_updateScrollInfoAfterLayoutTransaction.get(); }
+    bool markNeedingTransformUpdate(RenderLayerModelObject*, RenderLayerModelObject*);
+    Vector<WeakPtr<RenderLayerModelObject>> descendantsNeedingTransformUpdate(RenderLayerModelObject*);
```
`Vector<WeakPtr<RenderBox>> boxesNeedingTransformUpdateAfterContainerLayout(RenderBlock&)`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.h:
```diff
@@ -191,6 +195,7 @@ class LocalFrameViewLayoutContext {
     std::unique_ptr<Layout::LayoutTree> m_layoutTree;
     std::unique_ptr<Layout::LayoutState> m_layoutState;
     std::unique_ptr<UpdateScrollInfoAfterLayoutTransaction> m_updateScrollInfoAfterLayoutTransaction;
+    WeakHashMap<RenderLayerModelObject, Vector<WeakPtr<RenderLayerModelObject>>> m_containersWithDescendantsNeedingTransformUpdate;
```
`WeakHashMap<RenderBlock, Vector<WeakPtr<RenderBox>>>`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -627,6 +627,37 @@ void LocalFrameViewLayoutContext::popLayoutState()
+    if (!containingBlock)
+        return false;
+    auto it = m_containersWithDescendantsNeedingTransformUpdate.find(*containingBlock);
+    if (it != m_containersWithDescendantsNeedingTransformUpdate.end()) {
```
I still wonder if this can be written more efficiently with `ensure()`

smfr replied:
If not, you can still use the iterator returned by `find` for insertion.

nmoucht replied:
Ah right I misinterpreted the point of `ensure()`.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-022.html:
```diff
@@ -0,0 +1,35 @@
+      }
+
+      #target {
+
```
Can we remove this blank line?

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -287,6 +287,12 @@ TransformationMatrix* RenderLayerModelObject::layerTransform() const
 void RenderLayerModelObject::updateLayerTransform()
 {
+    if (is<RenderBox>(this) && style().offsetPath() && MotionPath::needsUpdateAfterContainingBlockLayout(*style().offsetPath())) {
+        if (auto cb = containingBlock()) {
```
```suggestion
        if (auto* containingBlock = this->containingBlock()) {
```

Let's avoid abbreviations per the webkit code style

nmoucht replied:
Fixed.

nmoucht committed:
[motion-path] Check if we are mid layout when setting containing block rect for ray
https://bugs.webkit.org/show_bug.cgi?id=260110
rdar://113780201

Reviewed by Simon Fraser.

On the first RenderLayer::updateTransform call the parent is mid layout, so it is
inorrect to query it about its rect size. If we have a motion path transform, set
that this RenderObject needs a transform update on the parent through the FrameView's
layout context, and have the parent call updateTransform after it has completed layout.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-019-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-019.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-020-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-020.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-021-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-021.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-022-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-022.html: Added.

Canonical link: https://commits.webkit.org/267479@main
